### PR TITLE
Introduce collective ops to ort inference build

### PR DESCRIPTION
**Pull request by souptc**
### Description
Introduce collective ops into onnxruntime inference build, including
1) AllReduce and AllGather schema in contrib op, controlled by USE_MPI flag
2) AllReduce and AllGather kernel in cuda EP, controlled by ORT_USE_NCCL flag

### Motivation and Context
Enable the collective ops in onnxruntime inference build so we have the ability to run distributed inference with multiple GPUs. 
The original ncclAllReduce ops in training build require quite complex configurations, which is not suitable for inference case, and it already broken. so we introduce a new implementation.

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/collective_defs.cc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          "Constrain to float, float16 and double tensors.")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        propagateShapeAndTypeFromFirstInput(ctx);
+      });
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(AllGather)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Attr("world_size",
+            "total world size that need to be gathered.",
```
Comment (gramalingam) [inline onnxruntime/core/graph/contrib_ops/collective_defs.cc:34]: Is "world-size" the commonly used word in these APIs? Otherwise, I would suggest a name like "num_processors" or "num_tensors" ,which I find a bit easier to understand. In any case, I would suggest updating the description to "Number of tensors to be gathered" or "Number of processors participating in the all-gather operation."

File: onnxruntime/core/graph/contrib_ops/collective_defs.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        assert(world_size >= static_cast<int64_t>(1));
+        // propagate type for output
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+        // propagate shape for output.
+        // output shape is [world_size * input_shape[0], ...]
+        auto output_type = ctx.getOutputType(0);
+        auto input_type = ctx.getInputType(0);
+        // all input/output should be tensor
+        assert(input_type->value_case() == output_type->value_case() && input_type->value_case() == TypeProto::kTensorType);
```
Comment (gramalingam) [inline onnxruntime/core/graph/contrib_ops/collective_defs.cc:54]: nit: The assert seems redundant given that the type is propagated just a few lines above and given the type-constraints above.

File: onnxruntime/test/python/onnxruntime_test_collective.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+print(f"!!!!!Running for rank {rank}")
+model = build_allreduce_model()
+ort_sess = ort.InferenceSession(
+    model.SerializeToString(),
+    providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
+    provider_options=[{"device_id": str(rank)}, {}],
+)
+input = np.ones((128, 128), dtype=np.float32)
+outputs = ort_sess.run(None, {"X": input})
+assert np.allclose(outputs[0], size * input)
```
Comment (gramalingam) [inline onnxruntime/test/python/onnxruntime_test_collective.py:35]: Does this work? I think `size * input` might multiple each value in `input` by `size` instead of concatenating.

File: onnxruntime/contrib_ops/cuda/cuda_contrib_kernels.cc [hunk 4]
```diff
@@ -133,124 +138,130 @@ KernelCreateInfo BuildKernelCreateInfo<void>() {
 
 Status RegisterCudaContribKernels(KernelRegistry& kernel_registry) {
   static const BuildKernelCreateInfoFn function_table[] = {
-      BuildKernelCreateInfo<void>,  // default entry to avoid the list become empty after ops-reducing
```
Comment (gramalingam) [inline onnxruntime/contrib_ops/cuda/cuda_contrib_kernels.cc:136]: Looks like lot of unrelated formatting changes.

Comment (souptc) [reply onnxruntime/contrib_ops/cuda/cuda_contrib_kernels.cc:136]: yes it is auto formatted.

File: onnxruntime/core/providers/cuda/cuda_provider_factory.cc [hunk 5]
```diff
@@ -28,7 +28,7 @@ using namespace onnxruntime;
 
 namespace onnxruntime {
 
-#if defined(USE_CUDA) && defined(ORT_USE_NCCL) && defined(USE_NCCL_P2P)
+#if defined(USE_CUDA) && defined(ORT_USE_NCCL) && defined(USE_NCCL_P2P) && defined(ENABLE_TRAINING)
```
[the remark this answers is not in the document]
Comment (souptc) [reply onnxruntime/core/providers/cuda/cuda_provider_factory.cc:31]: those component only available on training build. we need some cleanup later

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-distributed-test-ci-pipeline.yml [hunk 6]
```diff
@@ -53,3 +53,51 @@ jobs:
       condition: 'succeeded'
 
   - template: templates/clean-agent-build-directory-step.yml
+
+- job: Onnxruntime_Linux_GPU_Inference_Distributed_Test
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-distributed-test-ci-pipeline.yml:57]: You may put the two jobs in two stages, like what is in https://github.com/microsoft/onnxruntime/blob/main/tools/ci_build/github/azure-pipelines/post-merge-jobs.yml . Then when someone manually run this pipeline, they can choose to only run one of the jobs from Azure Devops' web UI.

Comment (souptc) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-ortmodule-distributed-test-ci-pipeline.yml:57]: thanks!

File: onnxruntime/test/python/onnxruntime_test_collective.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import unittest
```
Comment (pranavsharma) [inline onnxruntime/test/python/onnxruntime_test_collective.py:1]: Missing license header.

File: cmake/CMakeLists.txt [hunk 8]
```diff
@@ -1338,102 +1338,105 @@ if (onnxruntime_ENABLE_TRAINING)
...
-    endif()
+  add_subdirectory(tensorboard EXCLUDE_FROM_ALL)
+  list(APPEND onnxruntime_EXTERNAL_LIBRARIES tensorboard)
+endif()
 
-    find_package(MPI)
+if (UNIX)
+  if (EXISTS "${onnxruntime_MPI_HOME}")
+    set(MPI_HOME "${onnxruntime_MPI_HOME}")
+  elseif (EXISTS "/bert_ort/openmpi")
```
Comment (pranavsharma) [inline cmake/CMakeLists.txt:1348]: Do we need to hardcode these folder names in the else case?

File: cmake/CMakeLists.txt [hunk 9]
```diff
@@ -1338,102 +1338,105 @@ if (onnxruntime_ENABLE_TRAINING)
   add_compile_definitions(ENABLE_STRIDED_TENSORS)
   add_compile_definitions(ENABLE_TRAINING)
 
-  if (UNIX)
-    if (EXISTS "${onnxruntime_MPI_HOME}")
-      set(MPI_HOME "${onnxruntime_MPI_HOME}")
-    elseif (EXISTS "/bert_ort/openmpi")
-      set(MPI_HOME "/bert_ort/openmpi")
-    endif()
+  add_subdirectory(tensorboard EXCLUDE_FROM_ALL)
```
Comment (pranavsharma) [inline cmake/CMakeLists.txt:1341]: What is tensorboard used for? Can you add a comment?

File: onnxruntime/core/graph/contrib_ops/collective_defs.cc (hunk 1 above)
Comment (pranavsharma) [reply onnxruntime/core/graph/contrib_ops/collective_defs.cc:34]: Agree with Rama here. world_size is a word in the MPI vocab which is an implementation detail of this kernel. Can we rename this to something more generic and intuitive?

File: onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  } else if (type == DataTypeImpl::GetType<int64_t>()) {
+    return ncclInt64;
+  } else if (type == DataTypeImpl::GetType<MLFloat16>()) {
+    return ncclFloat16;
+  } else if (type == DataTypeImpl::GetType<float>()) {
+    return ncclFloat32;
+  } else if (type == DataTypeImpl::GetType<double>()) {
+    return ncclFloat64;
+  } else {
+    throw std::logic_error("Tensor type not supported in NCCL.");
```
Comment (pranavsharma) [inline onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc:24]: Can you use ORT_THROW_EX(std::invalid_argument) instead?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 11]
```diff
@@ -2708,6 +2708,10 @@ This op functions in much the same was as Dropout-11 and Dropout-13 do, execpt t
     RegisterNchwcSchemas();
   }
 #endif
+
+#ifdef USE_MPI
```
Comment (pranavsharma) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2712]: Should this be both USE_MPI && USE_NCCL given that all these collectives are implemented using NCCL?

File: cmake/CMakeLists.txt (hunk 8 above)
Comment (souptc) [reply cmake/CMakeLists.txt:1348]: it is inherit from the training team's setting, i don't want to break their build.

File: cmake/CMakeLists.txt (hunk 9 above)
Comment (souptc) [reply cmake/CMakeLists.txt:1341]: it is not my change, it existed in the current code, i just move the mpi related part out of training config.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 11 above)
Comment (souptc) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2712]: this is the schema part, not the kernels. we will need it one day another EP implement the collectives. So i don't think we should guard it with NCCL flag.

File: onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+  } else if (type == DataTypeImpl::GetType<int64_t>()) {
+    return ncclInt64;
+  } else if (type == DataTypeImpl::GetType<MLFloat16>()) {
+    return ncclFloat16;
+  } else if (type == DataTypeImpl::GetType<float>()) {
+    return ncclFloat32;
+  } else if (type == DataTypeImpl::GetType<double>()) {
+    return ncclFloat64;
+  } else {
+    ORT_THROW("Tensor type not supported in NCCL.");
```
Comment (pranavsharma) [inline onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc:28]: nit: Would be nice to print the type along with the error msg.

**Review by pranavsharma — APPROVED** [review]

**Review by jslhcl — DISMISSED** [review]

**Review by jchen351 — APPROVED** [review]
